File: lab3/gr9sub1/ЛУЦКИЙВАДИМ/src/pstat.py
#!/usr/bin/env python3
import os
import sys
import argparse
from pathlib import Path

class Colors:
    """Цвета для красивого вывода"""
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    END = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def read_proc_file(pid, filename):
    """Чтение файла из /proc с обработкой ошибок"""
    try:
        with open(f"/proc/{pid}/{filename}", "r") as f:
            return f.read()
    except FileNotFoundError:
        return None
    except PermissionError:
        return None

def parse_proc_stat(pid):
    """Парсинг /proc/<pid>/stat"""
    content = read_proc_file(pid, "stat")
    if not content:
        return None
    
    parts = content.split()
    return {
        'pid': parts[0],
        'comm': parts[1][1:-1],
        'state': parts[2],
        'ppid': parts[3],
        'utime': int(parts[13]),
        'stime': int(parts[14]),
        'num_threads': int(parts[19])
    }

def parse_proc_status(pid):
    """Парсинг /proc/<pid>/status"""
    content = read_proc_file(pid, "status")
    if not content:
        return None
    
    result = {}
    for line in content.split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            result[key.strip()] = value.strip()
    return result

def parse_proc_io(pid):
    """Парсинг /proc/<pid>/io"""
    content = read_proc_file(pid, "io")
    if not content:
        return None
    
    result = {}
    for line in content.split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            result[key.strip()] = int(value.strip())
    return result

def parse_smaps_data(pid):
    """Парсинг smaps_rollup или smaps для получения RssAnon и RssFile"""
    content = read_proc_file(pid, "smaps_rollup")
    if not content:
        content = read_proc_file(pid, "smaps")
    
    if not content:
        return None
    
    result = {}
    rss_anon = 0
    rss_file = 0
    rss_shmem = 0
    
    for line in content.split('\n'):
        if line.startswith('RssAnon:'):
            rss_anon = int(line.split()[1])
        elif line.startswith('RssFile:'):
            rss_file = int(line.split()[1])
        elif line.startswith('RssShmem:'):
            rss_shmem = int(line.split()[1])
    
    result['RssAnon'] = rss_anon
    result['RssFile'] = rss_file + rss_shmem
    return result

def format_bytes(size_bytes):
    """Форматирование байтов в человекочитаемый вид"""
    if size_bytes >= 1024 * 1024:
        return f"{size_bytes / (1024 * 1024):.2f} MiB"
    elif size_bytes >= 1024:
        return f"{size_bytes / 1024:.2f} KiB"
    else:
        return f"{size_bytes} B"

def format_kb_to_mib(kb_value):
    """Конвертация килобайтов в мегабайты"""
    try:
        if isinstance(kb_value, str):
            kb = int(kb_value.split()[0])
        else:
            kb = int(kb_value)
        return kb / 1024.0
    except (ValueError, TypeError, IndexError):
        return 0

def ticks_to_seconds(ticks, HZ=100):
    """Конвертация тиков в секунды"""
    return ticks / HZ

def print_section(title):
    """Печать заголовка секции"""
    print(f"\n{Colors.HEADER}{Colors.BOLD}{title}{Colors.END}")
    print("=" * 50)

def print_field(name, value, color=Colors.CYAN):
    """Печать поля с выравниванием"""
    print(f"{color}{name:<30}{Colors.END} {value}")

def main():
    parser = argparse.ArgumentParser(description='Утилита pstat для просмотра метрик процесса')
    parser.add_argument('pid', type=int, help='PID процесса')
    args = parser.parse_args()
    
    pid = args.pid
    
    # Проверяем существование процесса
    if not Path(f"/proc/{pid}").exists():
        print(f"{Colors.RED}Ошибка: Процесс с PID {pid} не найден{Colors.END}")
        sys.exit(1)
    
    # Собираем данные
    stat_data = parse_proc_stat(pid)
    status_data = parse_proc_status(pid)
    io_data = parse_proc_io(pid)
    smaps_data = parse_smaps_data(pid)
    
    if not stat_data:
        print(f"{Colors.RED}Не удалось прочитать данные процесса{Colors.END}")
        sys.exit(1)
    
    # Вычисляем производные значения
    HZ = os.sysconf(os.sysconf_names['SC_CLK_TCK'])
    cpu_time_sec = ticks_to_seconds(stat_data['utime'] + stat_data['stime'], HZ)
    
    # Получаем информацию о памяти
    vm_rss_mib = 0
    if status_data and 'VmRSS' in status_data:
        vm_rss_mib = format_kb_to_mib(status_data['VmRSS'])
    
    # Получаем RssAnon и RssFile
    rss_anon_mib = 0
    rss_file_mib = 0
    if smaps_data:
        rss_anon_mib = format_kb_to_mib(smaps_data.get('RssAnon', 0))
        rss_file_mib = format_kb_to_mib(smaps_data.get('RssFile', 0))
    
    # Извлекаем значения переключений контекста
    voluntary_switches = status_data.get('voluntary_ctxt_switches', 'N/A') if status_data else 'N/A'
    nonvoluntary_switches = status_data.get('nonvoluntary_ctxt_switches', 'N/A') if status_data else 'N/A'
    
    # Вывод результатов
    print(f"\n{Colors.BOLD}{Colors.HEADER}🔍 СТАТИСТИКА ПРОЦЕССА {pid} ({stat_data['comm']}){Colors.END}")
    print("=" * 60)
    
    print_section("📋 ОСНОВНАЯ ИНФОРМАЦИЯ")
    print_field("PID", f"{pid}")
    print_field("Имя процесса", stat_data['comm'])
    print_field("Родительский PID (PPid)", stat_data['ppid'])
    print_field("Количество потоков", stat_data['num_threads'])
    print_field("Состояние", stat_data['state'])
    
    print_section("⏱️  ПРОЦЕССОРНОЕ ВРЕМЯ")
    print_field("Общее время CPU", f"{cpu_time_sec:.2f} сек")
    print_field("Время в пользовательском режиме", f"{stat_data['utime']} тиков")
    print_field("Время в режиме ядра", f"{stat_data['stime']} тиков")
    print_field("Частота HZ системы", f"{HZ}")
    
    print_section("🔄 ПЕРЕКЛЮЧЕНИЯ КОНТЕКСТА")
    print_field("Добровольные переключения", voluntary_switches)
    print_field("Принудительные переключения", nonvoluntary_switches)
    
    print_section("💾 ПАМЯТЬ")
    print_field("Физическая память (VmRSS)", f"{vm_rss_mib:.2f} MiB")
    print_field("Анонимная память (RssAnon)", f"{rss_anon_mib:.2f} MiB")
    print_field("Файловая память (RssFile)", f"{rss_file_mib:.2f} MiB")
    
    print_section("📁 ВВОД-ВЫВОД")
    if io_data:
        print_field("Прочитано байт", format_bytes(io_data.get('read_bytes', 0)))
        print_field("Записано байт", format_bytes(io_data.get('write_bytes', 0)))
    else:
        print_field("Прочитано байт", "N/A")
        print_field("Записано байт", "N/A")
    
    # Дополнительная информация из statm
    statm_content = read_proc_file(pid, "statm")
    if statm_content:
        statm_parts = statm_content.split()
        print_section("📊 ДОПОЛНИТЕЛЬНАЯ ИНФОРМАЦИЯ")
        print_field("Всего страниц памяти", statm_parts[0])
        print_field("Страниц RSS", statm_parts[1])
        print_field("Размер страницы", "4 KiB (типично)")
        print_field("Общий размер памяти", f"{int(statm_parts[0]) * 4} KiB")

if __name__ == "__main__":
    main()

File: lab3/gr9sub1/ВАНЛИ_РУСЛАН/src/pstat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <time.h>

#define MAX_BUF 4096
#define HZ 100  // стандартное значение HZ для Linux

// Структура для хранения данных из /proc/<pid>/stat
typedef struct {
    int pid;
    char comm[256];
    char state;
    int ppid;
    unsigned long utime;
    unsigned long stime;
    long num_threads;
} proc_stat_t;

// Структура для данных из /proc/<pid>/status
typedef struct {
    unsigned long vm_rss;
    unsigned long rss_anon;
    unsigned long rss_file;
    unsigned long voluntary_ctxt_switches;
    unsigned long nonvoluntary_ctxt_switches;
} proc_status_t;

// Структура для данных из /proc/<pid>/io
typedef struct {
    unsigned long read_bytes;
    unsigned long write_bytes;
} proc_io_t;

// Функция для чтения файла в буфер
int read_proc_file(const char *path, char *buf, size_t buf_size) {
    int fd = open(path, O_RDONLY);
    if (fd == -1) {
        return -1;
    }
    
    ssize_t bytes_read = read(fd, buf, buf_size - 1);
    close(fd);
    
    if (bytes_read <= 0) {
        return -1;
    }
    
    buf[bytes_read] = '\0';
    return 0;
}

// Парсинг /proc/<pid>/stat
int parse_proc_stat(int pid, proc_stat_t *stat) {
    char path[64];
    char buf[MAX_BUF];
    
    snprintf(path, sizeof(path), "/proc/%d/stat", pid);
    if (read_proc_file(path, buf, sizeof(buf)) == -1) {
        return -1;
    }
    
    // Парсим основные поля согласно формату /proc/<pid>/stat
    char *token = strtok(buf, " ");
    int field = 1;
    
    while (token != NULL) {
        switch (field) {
            case 1: stat->pid = atoi(token); break;
            case 2: 
                // Убираем скобки вокруг имени команды
                if (token[0] == '(') {
                    strncpy(stat->comm, token + 1, sizeof(stat->comm) - 1);
                    size_t len = strlen(stat->comm);
                    if (len > 0 && stat->comm[len - 1] == ')') {
                        stat->comm[len - 1] = '\0';
                    }
                }
                break;
            case 3: stat->state = token[0]; break;
            case 4: stat->ppid = atoi(token); break;
            case 14: stat->utime = strtoul(token, NULL, 10); break;
            case 15: stat->stime = strtoul(token, NULL, 10); break;
            case 20: stat->num_threads = atol(token); break;
        }
        token = strtok(NULL, " ");
        field++;
    }
    
    return 0;
}

// Парсинг /proc/<pid>/status
int parse_proc_status(int pid, proc_status_t *status) {
    char path[64];
    char buf[MAX_BUF];
    
    snprintf(path, sizeof(path), "/proc/%d/status", pid);
    if (read_proc_file(path, buf, sizeof(buf)) == -1) {
        return -1;
    }
    
    char *line = strtok(buf, "\n");
    while (line != NULL) {
        if (strncmp(line, "VmRSS:", 6) == 0) {
            sscanf(line + 6, "%lu", &status->vm_rss);
        } else if (strncmp(line, "RssAnon:", 8) == 0) {
            sscanf(line + 8, "%lu", &status->rss_anon);
        } else if (strncmp(line, "RssFile:", 8) == 0) {
            sscanf(line + 8, "%lu", &status->rss_file);
        } else if (strncmp(line, "voluntary_ctxt_switches:", 24) == 0) {
            sscanf(line + 24, "%lu", &status->voluntary_ctxt_switches);
        } else if (strncmp(line, "nonvoluntary_ctxt_switches:", 27) == 0) {
            sscanf(line + 27, "%lu", &status->nonvoluntary_ctxt_switches);
        }
        line = strtok(NULL, "\n");
    }
    
    return 0;
}

// Парсинг /proc/<pid>/io
int parse_proc_io(int pid, proc_io_t *io) {
    char path[64];
    char buf[MAX_BUF];
    
    snprintf(path, sizeof(path), "/proc/%d/io", pid);
    if (read_proc_file(path, buf, sizeof(buf)) == -1) {
        return -1;
    }
    
    char *line = strtok(buf, "\n");
    while (line != NULL) {
        if (strncmp(line, "read_bytes:", 11) == 0) {
            sscanf(line + 11, "%lu", &io->read_bytes);
        } else if (strncmp(line, "write_bytes:", 12) == 0) {
            sscanf(line + 12, "%lu", &io->write_bytes);
        }
        line = strtok(NULL, "\n");
    }
    
    return 0;
}

// Форматирование размера в человеко-читаемый вид
void format_size(unsigned long kb, char *buf, size_t buf_size) {
    if (kb >= 1024) {
        snprintf(buf, buf_size, "%.2f MiB", kb / 1024.0);
    } else {
        snprintf(buf, buf_size, "%lu KiB", kb);
    }
}

// Основная функция
int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Использование: %s <pid>\n", argv[0]);
        return 1;
    }
    
    int pid = atoi(argv[1]);
    if (pid <= 0) {
        fprintf(stderr, "Ошибка: неверный PID\n");
        return 1;
    }
    
    proc_stat_t stat = {0};
    proc_status_t status = {0};
    proc_io_t io = {0};
    
    // Чтение и парсинг данных
    if (parse_proc_stat(pid, &stat) == -1) {
        fprintf(stderr, "Ошибка чтения /proc/%d/stat: %s\n", pid, strerror(errno));
        return 1;
    }
    
    if (parse_proc_status(pid, &status) == -1) {
        fprintf(stderr, "Ошибка чтения /proc/%d/status: %s\n", pid, strerror(errno));
        // Продолжаем выполнение, так как некоторые поля могут быть доступны
    }
    
    if (parse_proc_io(pid, &io) == -1) {
        fprintf(stderr, "Ошибка чтения /proc/%d/io: %s\n", pid, strerror(errno));
        // Продолжаем выполнение
    }
    
    // Вывод результатов
    printf("=== pstat для PID %d (%s) ===\n", pid, stat.comm);
    printf("PPid:                      %d\n", stat.ppid);
    printf("Threads:                   %ld\n", stat.num_threads);
    printf("State:                     %c\n", stat.state);
    
    // Время CPU
    double cpu_time_sec = (stat.utime + stat.stime) / (double)HZ;
    printf("CPU time (utime+stime):    %.2f сек\n", cpu_time_sec);
    printf("  utime:                   %lu тиков\n", stat.utime);
    printf("  stime:                   %lu тиков\n", stat.stime);
    
    // Переключения контекста
    printf("voluntary_ctxt_switches:   %lu\n", status.voluntary_ctxt_switches);
    printf("nonvoluntary_ctxt_switches:%lu\n", status.nonvoluntary_ctxt_switches);
    
    // Память
    char rss_buf[64], anon_buf[64], file_buf[64];
    format_size(status.vm_rss, rss_buf, sizeof(rss_buf));
    format_size(status.rss_anon, anon_buf, sizeof(anon_buf));
    format_size(status.rss_file, file_buf, sizeof(file_buf));
    
    printf("VmRSS:                     %s\n", rss_buf);
    printf("RssAnon:                   %s\n", anon_buf);
    printf("RssFile:                   %s\n", file_buf);
    
    // Ввод-вывод
    char read_buf[64], write_buf[64];
    format_size(io.read_bytes / 1024, read_buf, sizeof(read_buf));
    format_size(io.write_bytes / 1024, write_buf, sizeof(write_buf));
    
    printf("read_bytes:                %s\n", read_buf);
    printf("write_bytes:               %s\n", write_buf);
    
    return 0;
}